edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/metrics/AvaliabilityMetrics.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.Core.Metrics
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Runtime.CompilerServices;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Metrics;
+    using Microsoft.Extensions.Logging;
+
+    public interface IAvailabilityMetric
+    {
+        void ComputeAvailability(ModuleSet desired, ModuleSet current);
+    }
+
+    public class AvailabilityMetrics : IAvailabilityMetric
+    {
+        readonly IMetricsGauge running;
+        readonly IMetricsGauge expectedRunning;
+        readonly ISystemTime time;
+        readonly ILogger log = Logger.Factory.CreateLogger<Availability>();
+
+        readonly List<Availability> availabilities;
+        readonly Lazy<Availability> edgeAgent;
+
+        public AvailabilityMetrics(IMetricsProvider metricsProvider, ISystemTime time)
+        {
+            this.running = metricsProvider.CreateGauge(
+                "total_time_running_correctly_seconds",
+                "The amount of time the module was specified in the deployment and was in the running state",
+                new List<string> { "module_name" });
+
+            this.expectedRunning = metricsProvider.CreateGauge(
+                "total_time_expected_running_seconds",
+                "The amount of time the module was specified in the deployment",
+                new List<string> { "module_name" });
+
+            this.time = time;
+            this.availabilities = new List<Availability>();
+            this.edgeAgent = new Lazy<Availability>(() => new Availability("edgeAgent", this.CalculateEdgeAgentDowntime(), this.time));
+        }
+
+        public void ComputeAvailability(ModuleSet desired, ModuleSet current)
+        {
+            /* Get all modules that are not running but should be */
+            var down = new HashSet<string>(current.Modules.Values
```

Comment:
what expectation should be if a module in desired module set and desired status is running, but that module doesn't exist in current module set?

Reply:
edge agent will be shutting it down. this check happens before the agent tries to reconcile the states. If it fails to reconcile, that's not availability's problem, the reconciler should log that.

Reply:
Clarification: the first time a module is deployed (exists in desired but not actual), it will not count for availability at all.